core/store/src/trie/mem/arena/alloc.rs

```diff
@@ -0,0 +1,124 @@
+use super::{ArenaMemory, ArenaSliceMut};
```

Comment:
These are great questions and let me walk you through my thought process - feel free to poke holes here and see if we can find a simpler solution.

First, we desire to (1) allocate dynamically sized structures, and (2) reference things by only their pointer, not a slice (which is 8 bytes extra). To avoid unsafe code: because of (1), we can only allocate a byte array since Rust does not support creating a new struct that is dynamically sized. Because of (2), since we cannot safely reference a dynamically sized memory region in Rust without using a fat pointer, this means we cannot use native pointers and must resort to indices inside an array.

Now because we desire a giant byte array to use indices into, we could have allocated a Vec<u8> but the problem is we don't know how big of an allocation we want to have. And thus mmap. (To be fair, a hole can be punched here that we incrementally allocate more Vec<u8>'s, and have indices be (which vec, which byte) which we encode inside a 64 bit int. I'm not sure if that's any better in complexity, I'm guessing not).

So now we have an mmap, we need to be able to allocate memory regions inside the mmap to these dynamically sized objects. typed-arena doesn't work because we don't have a type to begin with, and likely no type would ever work because it would internally keep native pointers which we already said won't work. generational-arena would seem to suffer from the same problem because the Arena expects a type T, which we don't have. For alloc-no-stdlib, I must admit I don't fully understand it, but it seems that it returns something that contains native fat pointers?

Reply:
Based on what you’re saying, it’s indeed likely that `alloc-no-stdlib` would be the only reasonable fit. It does basically the same, I think, as your arena allocator.

It will definitely return [something that contains fat pointers](https://docs.rs/alloc-no-stdlib/latest/alloc_no_stdlib/struct.AllocatedStackMemory.html), but if you already know the length out-of-band, you can destructure it into a raw pointer and a length, and when the deallocation time comes you can rebuild the slice in order to pass it to `free_cell`.

This might increase the number of places where we have `unsafe` blocks, but I believe it’d reduce the number of places that need to be validated for bugless operation.

Does that make sense to you?

Reply:
Given that the arena allocator implementation is essentially a very simple size-class-based bump allocator, I feel that it is better than introducing extra unsafe code, and to me it's debatable if using alloc-no-stdlib makes the code any better because I still don't understand how it works :)

Reply:
`alloc-no-stdlib` is just an allocatod that provides the ability to allocate a slice.

While you’re right that your code technically does not have any `unsafe` block, I still feel like it is just as bad as actually-unsafe code: any bug in here would cause us to build the wrong tree, which in turn would lead to eg. funds being misallocated to someone else. Actually it might even be worse than an `unsafe`-based vulnerability, because it’s not unsafe so it would perfectly reproduce across all validators, and result in funds motions being validated by the chain.

That said, looking at https://github.com/dropbox/rust-alloc-no-stdlib/blob/master/src/stack_allocator.rs#L34-L100 it doesn’t look like it’d be significantly better than your implementation, as it seems subject to fragmentation at a quick glance, whereas yours seems more reproducible :) I’m reviewing it properly then!